src/main/java/net/minecraftforge/common/util/MutableHashedLinkedList.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) Forge Development LLC and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+package net.minecraftforge.common.util;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+import org.apache.commons.lang3.mutable.MutableObject;
+import org.jetbrains.annotations.Nullable;
+
+import it.unimi.dsi.fastutil.Hash.Strategy;
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenCustomHashMap;
+
+public class MutableHashedLinkedList<K, V> implements Iterable<Map.Entry<K, V>>
+{
+    private final Map<K, Entry> entries;
+    private final MergeFunction<K, V> merge;
+    private final MutableObject<Entry> head = new MutableObject<>();
+    private final MutableObject<Entry> last = new MutableObject<>();
+
+    public MutableHashedLinkedList(Strategy<? super K> strategy) {
+        this(strategy, (k, v1, v2) -> v2);
+    }
+
+    public MutableHashedLinkedList(Strategy<? super K> strategy, MergeFunction<K, V> merge) {
+        this.entries = new Object2ObjectOpenCustomHashMap<>(strategy);
+        this.merge = merge;
+    }
+
+    public V put(K key, V value) {
```

Comment:
This method is missing a `@Nullable` annotation. Here are my suggested javadocs for this method (including the annotation):
```java
    /**
     * Inserts the mapping with the specified key and value pair.
     *
     * <p>If there is a mapping already associated with this key, then the previous value and the specified (new) value
     * are merged according to this collection's merge function, and the position of the entry is not modified. If there
     * is no such mapping, then the key-value mapping is inserted at the end of this collection.</p>
     *
     * @param key key to be inserted
     * @param value (new) value to be associated with the key
     * @return the previous value associated with the specified key, or {@code null} if there was no mapping for the key
     */
    @Nullable
```